sound/soc/sof/intel/debug.c

```diff
@@ -0,0 +1,287 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Cezary Rojewski <cezary.rojewski@intel.com>
+//
+
+#include <linux/debugfs.h>
+#include "../sof-priv.h"
+#include "probe.h"
+#include "compress.h"
+
+/**
+ * strsplit_u32 - Split string into sequence of u32 tokens
+ * @buf:	String to split into tokens.
+ * @delim:	String containing delimiter characters.
+ * @tkns:	Returned u32 sequence pointer.
+ * @num_tkns:	Returned number of tokens obtained.
+ */
+static int
+strsplit_u32(char **buf, const char *delim, u32 **tkns, size_t *num_tkns)
+{
+	char *s;
+	u32 *data, *tmp;
+	size_t count = 0;
+	size_t max_count = 32;
+	int ret = 0;
+
+	*tkns = NULL;
+	*num_tkns = 0;
+	data = kcalloc(max_count, sizeof(*data), GFP_KERNEL);
+	if (!data)
+		return -ENOMEM;
+
+	while ((s = strsep(buf, delim)) != NULL) {
+		ret = kstrtouint(s, 0, (data + count));
+		if (ret)
+			goto exit;
+		if (++count >= max_count) {
+			max_count *= 2;
+			tmp = kcalloc(max_count, sizeof(*data), GFP_KERNEL);
+			if (!tmp) {
+				ret = -ENOMEM;
+				goto exit;
+			}
+
+			memcpy(tmp, data, count * sizeof(*data));
+			kfree(data);
+			data = tmp;
+		}
+	}
+
+	if (!count)
+		goto exit;
+	*tkns = kcalloc(count, sizeof(*data), GFP_KERNEL);
+	if (*tkns == NULL) {
+		ret = -ENOMEM;
+		goto exit;
+	}
+	memcpy(*tkns, data, count * sizeof(*data));
+	*num_tkns = count;
+
+exit:
+	kfree(data);
```

Comment:
looks like you have memory leaks with tmp allocated on line 44?
This code looks error-prone and too custom....

Reply:
Don't see the leak, really. But your comment made me replace the tmp assignment with krealloc rather than kcalloc.

The "exit" path covers for mem leaks, unless I'm really blind.
About the customness - this function is intended for more than just probes. In cAVS I'm using it for traces too. In general I've written it in such a way that everyone could make use of it..
Such API should be part of common kernel code - right now only the most basic functions are and everyone is parsing string of DWORDs by themselves.